lib/rack/attack.rb

```diff
@@ -128,6 +146,8 @@ def call(env)
         configuration.tracked?(request)
         @app.call(env)
       end
+    rescue StandardError => error
+      self.class.allow_error?(error) ? @app.call(request.env) : raise(error)
```

Comment:
So, here we are rescuing `StandardError` so we can allow strings in the `allowed_errors` configuration, right? Is there any reason why to not do this?

```ruby
rescue allowed_errors => error
  @app.call(request.env)
end
```

because I prefer the simplicity of this alternative if possible

Reply:
I did it this way so we can define the `allowed_errors` as strings, and therefore not have a dependency on `Dalli` or `Redis` constants being loaded. Note the `allow_error?` method is as follows:

```ruby
      def allow_error?(error)
        allowed_errors&.any? do |ignored_error|
          case ignored_error
          when String then error.class.ancestors.any? {|a| a.name == ignored_error }
          else error.is_a?(ignored_error)
          end
        end
      end
```

Reply:
>not have a dependency on Dalli or Redis constants being loaded

Agree on this.

Not really sold on the `allow_error?` method, I'd prefer to use Ruby's exception handling instead, while finding other way to solve the need to not load `Dalli`/`Redis` constants.

I know you didn't like it before, but one way could be the `Rack::Attack::CacheError` class, but there could be others, for example having this class wrap and re-raise the original error, maybe? Or maybe having the defaults defined by each proxy class which also checks for `defined?(Dalli)` [as they already do in other places](https://github.com/rack/rack-attack/blob/main/lib/rack/attack/store_proxy/dalli_proxy.rb#L10)?

I think this point is really important and we need to take the time to brainstorm/discuss it properly.

Reply:
>In order to invoke rescue *allowed_errors, the allowed_errors variable must contain constants, not strings.

I know that. But why do we need to also allow `String`s? If supporting only constants makes the code simpler, then that works for me.

Reply:
As I've explained above, we cannot simply support only constants here. We must assume that `Rack::Attack` might be loaded before `Dalli` and `Redis`.

If we check for `defined?(Dalli)` at the time `Rack::Attack` loads, we may end up not setting those errors in the list. Unless we do something like dynamically converting strings to constants, which as I've illustrated above adds complexity.

If you would like to make a code suggestion on how this might be implemented, by all means please do. Otherwise, let's go forward with what I have here.

Reply:
one option would be to do:

```
def allowed_errors
  @allowed_errors || default_allowed_errors
end

def default_allowed_errors
   arr = []
   arr = Dalli::DalliError if defined?(Dalli)
   arr = Redis::BaseError if defined?(Redis)
   arr
end
```

I do not like this however, because:
1. doing `Rack::Attack.allowed_errors += [MyError]` will pin the value e.g. if Redis hasn't loaded yet
2. we are doing a dynamic build on every Rack::Attack call
3. the string approach is lower complexity/fewer LOC
4. as a bonus, the string approach also allows the user to define errors as strings for which consts may not have loaded yet

Reply:
I think there are other ways I said above

>I know you didn't like it before, but one way could be the Rack::Attack::CacheError class, but there could be others, for example having this class wrap and re-raise the original error, maybe? Or maybe having the defaults defined by each proxy class which also checks for defined?(Dalli) [as they already do in other places](https://github.com/rack/rack-attack/blob/main/lib/rack/attack/store_proxy/dalli_proxy.rb#L10)?

Some options:
1. `Rack::Attack::CacheError` as shown in https://github.com/rack/rack-attack/compare/main...santib:rack-attack:system-hardening-refactor#diff-9114718ab0fa2d856234c04d73b07c3758a6aaab1dbfbe87401f4eb4e358f127R134
2. `Rack::Attack::CacheError` as a wrapper for the original error, so we don't lose it
3. Having the store proxies dynamically add to an array of `allowed_errors` when they are invoked (to prevent the issue you are mentioning about load order)

I personally would prefer to go with 1. or 2., but I'd like to hear @grzuy thoughts on this.

Reply:
For reference, there is precedent in other Ruby libs to reference potentially unloaded classes with strings, such as Rails where you can see this:

```ruby
belongs_to :manager, class_name: "Employee"
```

Reply:
Last thing: this is a moot point. Please see these line in the intended final state of this series of PRs:

https://github.com/rack/rack-attack/pull/577/files#diff-9114718ab0fa2d856234c04d73b07c3758a6aaab1dbfbe87401f4eb4e358f127R218-R225

As you can see, I'm adding an additional error_handler here.

Reply:
> Rack::Attack::CacheError

I would really prefer that we do not obscure the errors coming from Redis/Dalli itself. It makes it a lot harder to work with and understand the reasons why Redis/Dalli are failing.

Reply:
>>Rack::Attack::CacheError
>
>I would really prefer that we do not obscure the errors coming from Redis/Dalli itself. It makes it harder to understand the reasons why Redis/Dalli are failing.

it could be just a wrapper, and then pass the original error to the application

Reply:
Let's defer this discussion until I get my other PRs merged.